8303762: [vectorapi] Intrinsification of Vector.slice

merykitty opened the pull request:
`Vector::slice` is a method at the top-level class of the Vector API that concatenates the 2 inputs into an intermediate composite and extracts a window equal to the size of the inputs into the result. It is used in vector conversion methods where the part number is not 0 to slice the parts to the correct positions. Slicing is also used in text processing such as utf8 and utf16 validation. x86 starting from SSSE3 has `palignr` which does vector slicing very efficiently. As a result, I think it is beneficial to add a C2 node for this operation as well as intrinsify `Vector::slice` method.

A slice is currently implemented as `v2.rearrange(iota).blend(v1.rearrange(iota), blendMask)` which requires preparation of the index vector and the blending mask. Even with the preparations being hoisted out of the loops, microbenchmarks show improvement using the slice instrinsics. Some have tremendous increases in throughput due to the limitation that a mask of length 2 cannot currently be intrinsified, leading to falling back to the Java implementations.

Please take a look and have some reviews. Thank you very much.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303762](https://bugs.openjdk.org/browse/JDK-8303762): [vectorapi] Intrinsification of Vector.slice (**Enhancement** - P4)


### Reviewers
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to [65409f13](https://git.openjdk.org/jdk/pull/12909/files/65409f136e506b345aeb699614ed1613e80e8de2)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12909/head:pull/12909` \
`$ git checkout pull/12909`

Update a local copy of the PR: \
`$ git checkout pull/12909` \
`$ git pull https://git.openjdk.org/jdk.git pull/12909/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12909`

View PR using the GUI difftool: \
`$ git pr show -t 12909`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12909.diff">https://git.openjdk.org/jdk/pull/12909.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12909#issuecomment-1458642895)

merykitty committed:
vector slice intrinsics

merykitty committed:
add jmh

merykitty committed:
optimise 64B

merykitty committed:
aesthetic

merykitty committed:
sse2, increase warmup

merykitty commented:
/label hotspot-compiler

PaulSandoz commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java:
```diff
@@ -2279,14 +2279,21 @@ public final ByteVector blend(long e,
+        return VectorSupport.slice(
+            getClass(), byte.class, length(),
+            this, that, origin,
+            new VectorSliceOp<ByteVector>() {
```
Change from inner class to lambda expression?

merykitty replied:
We still need this method to be inlined and I don't know if there is a way to annotate the lambda function.

PaulSandoz replied:
Yes, i wondered about the inline and how important it might be. You want the fallback to inline so as not to perturb platforms without the intrinsic. Can you add a comment on the anon class?

PaulSandoz commented on test/hotspot/jtreg/compiler/vectorapi/TestVectorSlice.java:
```diff
@@ -0,0 +1,918 @@
+
+        int length = 8;
+        testB64(dst, src1, src2);
+        for (int i = 0; i <= length; i++) {
+            for (int j = 0; j < length; j++) {
+                int offset = i + j;
+                byte expected;
+                if (offset < length) {
+                    expected = (byte)offset;
+                } else {
+                    expected = (byte)(offset + diff - length);
+                }
+                Asserts.assertEquals(expected, dst[i][j]);
+            }
+        }
```
It should be possible to factor out this code into something like this:
```
  assertOffsets(length, (expected, i, j) -> Assert.assertEquals((byte)expected, dst[i][j])
```

merykitty replied:
Fixed.

PaulSandoz commented on test/hotspot/jtreg/compiler/vectorapi/TestVectorSlice.java:
```diff
@@ -0,0 +1,918 @@
+        }
+
+        length = 16;
+        testB128(dst, src1, src2);
```
Should `dst` be zeroed before the next call?

merykitty replied:
Fixed, I just allocate another array.

PaulSandoz commented on test/jdk/jdk/incubator/vector/templates/Kernel-Slice-bop-const.template:
```diff
@@ -0,0 +1,11 @@
+        $type$[] a = fa.apply(SPECIES.length());
```
Forgot to commit the updated unit tests?

merykitty replied:
This is for the microbenchmarks generated in the panama-vector repo only.
Thanks a lot.

merykitty committed:
address reviews

PaulSandoz reviewed (approved):
Java changes look good. The HotSpot code looks well structured but i will let others comment on the specifics.

merykitty committed:
add comments explaining anonymous classes

merykitty commented:
@PaulSandoz Thanks for your review, I have added a comment explaining the rationales behind the anonymous class usage.

merykitty committed:
Merge branch 'master' into sliceIntrinsics

merykitty committed:
instruction asserts

PaulSandoz commented:
With the latest PR I am observing failures with debug builds for test compiler/vectorapi/TestVectorSlice.java on both AVX512 machines and aarch64 machines.

On AVX512 machines the test fails with JVM args `-XX:UseAVX=3` and `-XX:UseAVX=3 -XX:+UnlockDiagnosticVMOptions -XX:+UseKNLSetting` and results in a test assertion failure e.g.,
```
Caused by: java.lang.RuntimeException: assertEquals: expected 70 to equal 0
	at jdk.test.lib.Asserts.fail(Asserts.java:594)
	at jdk.test.lib.Asserts.assertEquals(Asserts.java:205)
	at jdk.test.lib.Asserts.assertEquals(Asserts.java:189)
	at compiler.vectorapi.TestVectorSlice.lambda$testInts$2(TestVectorSlice.java:163)
	at compiler.vectorapi.TestVectorSlice.testInts(TestVectorSlice.java:181)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 7 more
```

CPU flags are:
```
fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant tsc arch perfmon rep good nopl xtopology cpuid tsc known freq pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4 1 sse4 2 x2apic movbe popcnt tsc deadline timer aes xsave avx f16c rdrand hypervisor lahf lm abm 3dnowprefetch cpuid fault invpcid single ssbd ibrs ibpb stibp ibrs enhanced tpr shadow vnmi flexpriority ept vpid ept ad fsgsbase tsc adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves nt good wbnoinvd arat avx512vbmi umip pku ospke avx512 vbmi2 gfni vaes vpclmulqdq avx512 vnni avx512 bitalg avx512 vpopcntdq la57 rdpid md clear arch capabilities
```

On aarch64 there is an IR rule failure.

XiaohongGong commented on src/hotspot/share/opto/vectorIntrinsics.cpp:
```diff
@@ -1893,6 +1893,74 @@ bool LibraryCallKit::inline_vector_blend() {
+                    NodeClassNames[argument(5)->Opcode()]);
+    }
+    return false; // not enough info for intrinsification
+  }
+  if (!is_klass_initialized(vector_klass)) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** klass argument not initialized");
+    }
+    return false;
+  }
+  ciType* elem_type = elem_klass->const_oop()->as_instance()->java_mirror_type();
+  if (!elem_type->is_primitive_type()) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** not a primitive bt=%d", elem_type->basic_type());
+    }
+    return false; // should be primitive type
+  }
+  BasicType elem_bt = elem_type->basic_type();
```
Code style: It's better to add a blank line between different blocks.

XiaohongGong commented on src/hotspot/share/opto/vectorIntrinsics.cpp:
```diff
@@ -1893,6 +1893,74 @@ bool LibraryCallKit::inline_vector_blend() {
+
+  if (!arch_supports_vector(Op_VectorSlice, num_elem, elem_bt, VecMaskNotUsed)) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** not supported: arity=2 op=slice vlen=%d etype=%s ismask=notused",
+                    num_elem, type2name(elem_bt));
```
`ismask=notused` could be removed. We used `ismask` in other intrinsics to print whether it is a vector mask operation instead of vector class.

XiaohongGong commented on src/hotspot/share/opto/vectorIntrinsics.cpp:
```diff
@@ -1893,6 +1893,74 @@ bool LibraryCallKit::inline_vector_blend() {
+
+  Node* v1   = unbox_vector(argument(3), vbox_type, elem_bt, num_elem);
+  Node* v2   = unbox_vector(argument(4), vbox_type, elem_bt, num_elem);
+  Node* o    = argument(5);
+
+  if (v1 == NULL || v2 == NULL) {
+    return false; // operand unboxing failed
+  }
```
Suggest to reorder line-1950 and the if-statement in line-1952. And then we doesn't need too more spaces in the variable definition `Node* v1 = unbox_vector(xxx)`. Besides, could we rename variable `o` to `index` or `origin` ? I know you'v used `origin` at the begin, maybe we can rename it to `origin_type`. I see the similari name style in `inline_vector_frombits_coerced`.

merykitty committed:
add identity, fix flags

merykitty committed:
style

merykitty commented:
@PaulSandoz I have fixed the error in AVX512 and added feature predicates to not do IR check on AArch64
@XiaohongGong Thanks for your reviews, I have addressed them

PaulSandoz commented:
Tier 1-3 tests now pass.

XiaohongGong commented on test/hotspot/jtreg/compiler/vectorapi/TestVectorSlice.java:
```diff
@@ -0,0 +1,917 @@
+
+    @Test
+    @IR(counts = {IRNode.VECTOR_SLICE, "17"})
+    static void testB128(byte[][] dst, byte[] src1, byte[] src2) {
+        var species = ByteVector.SPECIES_128;
```
Suggest to define the species as a "`private static final`" field of this test class. It may make the intrinsification fail if the species is not a constant to the compiler.

merykitty replied:
This local is final and is loaded from a `static final` field so it should be equivalent to referring to `ByteVector.SPECIES_128` directly

XiaohongGong commented on test/hotspot/jtreg/compiler/vectorapi/TestVectorSlice.java:
```diff
@@ -324,7 +324,7 @@ static void testDoubles() {
 
     @Test
-    @IR(counts = {IRNode.VECTOR_SLICE, "9"})
+    @IR(counts = {IRNode.VECTOR_SLICE, "7"}, applyIfCPUFeature = {"sse2", "true"})
```
How about separating the special cases (i.e. origin is `0/VLENGTH`), and use the `FailOn` check instead on them? Tests are more accurate.

e1iu commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -1988,6 +1988,22 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+  // (VectorSlice X Y VLENGTH) => Y
+  if (origin->is_con(0)) {
+    return in(1);
+  } else if (origin->is_con(Matcher::vector_length(this))) {
```
If they were the same, length() looks simple.
```suggestion
  } else if (origin->is_con(length())) {
```

e1iu commented on src/java.base/share/classes/jdk/internal/vm/vector/VectorSupport.java:
```diff
@@ -627,6 +628,24 @@ V blend(Class<? extends V> vClass, Class<M> mClass, Class<E> eClass,
+        V apply(V v1, V v2, int origin);
+    }
+
+    @ForceInline
```
May I ask why `forceInline` here is necessary?

jatin-bhateja commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -7935,6 +7934,183 @@ instruct evblendvp64_masked(vec dst, vec src1, vec src2, kReg mask) %{
+%}
+
+instruct evsliceILFD(vec dst, vec src1, vec src2, immI origin) %{
+  predicate((type2aelembytes(Matcher::vector_element_basic_type(n)) * n->in(2)->get_int()) % 4U == 0 &&
+            (Matcher::vector_length_in_bytes(n) == 64 ||
+             (Matcher::vector_length_in_bytes(n) == 32 &&
+              VM_Version::supports_avx512vl())));
```
Same as above.

jatin-bhateja commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -7935,6 +7934,183 @@ instruct evblendvp64_masked(vec dst, vec src1, vec src2, kReg mask) %{
+  format %{ "vector_slice $dst, $src, $origin\t! using $xtmp as TEMP" %}
+  ins_encode %{
+    BasicType bt = Matcher::vector_element_basic_type(this);
+    int shift_count = $origin$$constant * type2aelembytes(bt);
+    __ psrldq($dst$$XMMRegister, shift_count);
+    __ movdqu($xtmp$$XMMRegister, $src$$XMMRegister);
+    __ pslldq($xtmp$$XMMRegister, 16 - shift_count);
+    __ por($dst$$XMMRegister, $xtmp$$XMMRegister);
```
Move to macro assembly routine.

jatin-bhateja commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -7935,6 +7934,183 @@ instruct evblendvp64_masked(vec dst, vec src1, vec src2, kReg mask) %{
+  format %{ "vector_slice $dst, $src1, $src2, $origin" %}
+  ins_encode %{
+    int vlen = Matcher::vector_length_in_bytes(this);
+    BasicType bt = Matcher::vector_element_basic_type(this);
+    int shift_count = $origin$$constant * type2aelembytes(bt);
+    if (vlen == 16) {
+      __ vpalignr($dst$$XMMRegister, $src2$$XMMRegister, $src1$$XMMRegister, shift_count, Assembler::AVX_128bit);
+    } else {
+      if (vlen == 8) {
+        __ vpunpcklqdq($dst$$XMMRegister, $src1$$XMMRegister, $src2$$XMMRegister, Assembler::AVX_128bit);
+      } else {
+        assert(vlen == 4, "");
+        __ vpunpckldq($dst$$XMMRegister, $src1$$XMMRegister, $src2$$XMMRegister, Assembler::AVX_128bit);
+      }
+      __ vpsrldq($dst$$XMMRegister, $dst$$XMMRegister, shift_count, Assembler::AVX_128bit);
+    }
```
Move to macro assembly routine.
